backup/metadata.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package backup
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/version"
+
+	"github.com/juju/juju-restore/core"
+)
+
+func readMetadataJSON(path string) (core.BackupMetadata, error) {
+	source, err := os.Open(path)
+	if err != nil {
+		return core.BackupMetadata{}, errors.Trace(err)
+	}
+	defer source.Close()
+	data, err := ioutil.ReadAll(source)
+
+	// Try the current version and check the FormatVersion first.
+	var target flatMetadata
+	err = json.Unmarshal(data, &target)
+	if err != nil {
+		return core.BackupMetadata{}, errors.Annotate(err, "unmarshalling v1 metadata")
+	}
+
+	if target.FormatVersion == 1 {
+		return flatToBackupMetadata(target), nil
+	}
+
+	// No FormatVersion set - it must be a version 0 structure
+	// instead.
+	var targetV0 flatMetadataV0
+	err = json.Unmarshal(data, &targetV0)
+	if err != nil {
+		return core.BackupMetadata{}, errors.Annotate(err, "unmarshalling v0 metadata")
+	}
+	return flatV0ToBackupMetadata(targetV0), nil
+}
+
+// Duplicating the flat metadata formats from the juju codebase for
```

Comment:
Yes.. I wonder where we can share it?... juju/description? Or just have it defined in here and have juju/juju-restore as an import in juju/juju? Or having a separate dedicated formats repo?...

Worth pondering :)